dts: ti: cc13xx_cc26xx: devicetree sysclk alignment

fgrandel opened the pull request:
This change introduces the "_rtc_timer" suffix for the system tick timer driver "compatible" property and aligns naming conventions with the actual CC13/26xx SoC series product policy.

This frees up the "_rtc" namespace to introduce additional APIs based on the same peripheral in the future:

```
rtc: rtc@... {
  compatible = "ti,cc13xx-cc26xx-rtc";
  ...

  timer {
    compatible = "ti,cc13xx-cc26xx-rtc-timer";
    ...
  };

  counter {
    compatible = "ti,cc13xx-cc26xx-rtc-counter";
    ...
  };

  pps {
    compatible = "ti,cc13xx-cc26xx-rtc-pps";
    ...
  };
};
```

Or alternatively an MFD pattern with similar requirements.

To enable this change, some duplicate devicetree nodes were removed and inconsistencies in the naming and include hierarchy of the cc13xx_cc26xx devicetree and RTC system clock driver fixed.

fgrandel commented:
@cfriedt @niflostancu You'll probably also want to review this.

cfriedt reviewed (commented):
It's possible that this may be usable as a RTC in spite of the fact that it's missing separate year, month, day of month, etc registers. 

In particular, the fact that it has support for compensation indicates to me that it might actually work.

I wouldn't mind hearing @vaishnavachath's thoughts here.

fgrandel commented:
> It's possible that this may be usable as a RTC in spite of the fact that it's missing separate year, month, day of month, etc registers.

@cfriedt True, the hardware peripheral has the capability for system_timer.h, rtc.h, counter.h and [PPS](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731) for clock synchronization.

Got some help from Brix on Discord, I guess the correct solution would then be aon_rtc: rtc@... following the "peripheral-name-from-datasheet: generic-function@address" pattern? The same is probably true for the compat-String which was fine as it represents hardware, not the driver software. Will update the PR accordingly.

fgrandel commented:
@cfriedt Update: As [only one driver can be attached per dt node](https://github.com/zephyrproject-rtos/zephyr/issues/48934) using the _rtc_timer suffix for the compatible property is correct as we'd clutter the namespace otherwise. Do you agree?

vaishnavachath commented on dts/arm/ti/cc13xx_cc26xx.dtsi:
```diff
@@ -123,11 +123,12 @@
+		/* The RTC peripheral backs the kernel system clock and tick timer. */
 		rtc: rtc@40092000 {
-			compatible = "ti,cc13xx-cc26xx-rtc";
+			compatible = "ti,cc13xx-cc26xx-rtc-timer";
```
@fgrandel Thank you for the PR, when counter or PPS support is added how would the node change, I am trying to correlate the description in the PR and the changes, here only the compatible seems to be changed?

fgrandel replied:
@vaishnavachath Not sure whether I understand your question... When something like PPS is upstreamed, then it will probably initially look like this:

```
rtc: rtc@40092000 {
  compatible = "ti,cc13xx-cc26xx-rtc";
  reg = <0x40092000 0x1000>;
  interrupts = <4 0>;  /* interrupt #20 = 4 + 16 */

  timer {
    compatible = "ti,cc13xx-cc26xx-rtc-timer";
    status = "okay"; /* the system clock timer is mandatory */
  };

  pps {
    compatible = "ti,cc13xx-cc26xx-rtc-pps";
    status = "disabled"
  };
};
```

This is a quite established pattern for flexible peripherals with multiple uses/drivers, see e.g. this one:

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/f0/stm32f0.dtsi#L249

```
		timers3: timers@40000400 {
			compatible = "st,stm32-timers";
			reg = <0x40000400 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00000002>;
			resets = <&rctl STM32_RESET(APB1, 1U)>;
			interrupts = <16 0>;
			interrupt-names = "global";
			st,prescaler = <0>;
			status = "disabled";

			pwm {
				compatible = "st,stm32-pwm";
				status = "disabled";
				#pwm-cells = <3>;
			};

			counter {
				compatible = "st,stm32-counter";
				status = "disabled";
			};
		};
```

Also see this discussion re MFD devices that goes into quite some detail and links to even more detailed precursors: https://github.com/zephyrproject-rtos/zephyr/issues/48934

vaishnavachath replied:
@fgrandel , Thanks for clarifying, in #48934 the scheme mentioned is:
```
 bus {
       parent {
          child1 {
          }; 
          child2 {
          };
          child3 {
          };
       };
   };
```
Here you have updated the child1, but I did not see the parent node changes in the diff, in the above example also you mentioned (added some comments on where I am slightly confused):
```
rtc: rtc@40092000 {
  compatible = "ti,cc13xx-cc26xx-rtc"; /* parent device : I don't see this in your diff */
  reg = <0x40092000 0x1000>;
  interrupts = <4 0>;  /* interrupt #20 = 4 + 16 */

  timer {
    compatible = "ti,cc13xx-cc26xx-rtc-timer"; /* child device: in your diff when you changed the compatible this is the parent device now?? */
    status = "okay"; /* the system clock timer is mandatory */
  };

  pps {
    compatible = "ti,cc13xx-cc26xx-rtc-pps";
    status = "disabled"
  };
};
```

fgrandel replied:
(Cleared this up over Discord) The misunderstanding here was that my PR description and commit comment didn't make it clear enough that this was a preparative change that was meant to ensure that names would not have to be changed (and backwards compat broken) in the future but that the hierarchical devicetree would not be introduced until actually required.

vaishnavachath replied:
@fgrandel Thanks for the clarification in chat, looking at the commit message/PR description, I assumed that you are adding the parent node as well that was the confusion.

vaishnavachath commented on dts/arm/ti/cc13xx_cc26xx.dtsi:
```diff
@@ -123,11 +123,12 @@
 			reg = <0x40092000 0x1000>;
 			interrupts = <4 0>;  /* interrupt #20 = 4 + 16 */
-			status = "disabled";
+			status = "okay"; /* the system clock timer is mandatory */
```
Do we really need to default enable this at an SoC family level? I am not sure if this is required for a very minimal system if not I think it would be preferred to keep the existing scheme, so that users using different hardware platforms are not forced with defaults that they are not aware/needed for a minimal system.

fgrandel replied:
@vaishnavachath The system clock/timer driver is required in all cases that I'm aware of, as basic kernel thread scheduling depends on it. AFAIK several SoCs don't even bother to expose the system clock driver via DT and just mandatorily add it in the background.

In the rare case where s.o. actually wants to disable this (e.g. to provide an alternative sysclock driver), an overlay can still be created that disables the (sane) default. Someone with such an expert requirement will probably know how to do that. IMO it's much more probable, that less knowledgeable users will forget to add this peripheral for their custom boards which will provoke strange error messages they don't understand.

But maybe I overlook something? Let me know...

vaishnavachath reviewed (changes requested):
@fgrandel can we split the changes to two separate commits? one for the binding/DT update and another for the rename/include fixes?

Commit:
soc: ti: cc13/26xx: clean up include hierarchy

Removes duplicate code and inconsistencies in the naming of the
cc13xx_cc26xx devicetree and RTC driver hierarchy and alignes it with
the actual TI product series naming hierarchy.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
dts: ti: cc13xx_cc26xx: devicetree sysclk alignment

This change introduces the "_rtc_timer" suffix for the system tick timer
driver "compatible" property and aligns naming conventions with the
actual CC13/26xx SoC series product policy.

This frees up the "_rtc" namespace to introduce additional APIs based on
the same peripheral in the future (not part of this PR):

rtc: rtc@... {
  compatible = "ti,cc13xx-cc26xx-rtc";
  ...

  timer {
    compatible = "ti,cc13xx-cc26xx-rtc-timer";
    ...
  };

  counter {
    compatible = "ti,cc13xx-cc26xx-rtc-counter";
    ...
  };

  pps {
    compatible = "ti,cc13xx-cc26xx-rtc-pps";
    ...
  };
};

Or alternatively an MFD pattern with similar requirements.

Fixing the namespacing now makes sense standalone as it reduces the
chance of custom drivers being broken in the future.

Redundant extension of the mandatory system clock devicetree node is
replaced with a single `status = "okay"` which seems to be the more
sensible default to avoid user error when defining custom boards.
Knowledgeable users can still override this if really needed.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
> an we split the changes to two separate commits? one for the binding/DT update and another for the rename/include fixes?

@vaishnavachath Yes, of course. Makes sense.

fgrandel commented:
Echo client/server regression test result:

Interface 0x20001e48 (IEEE 802.15.4) [1]
========================================
IPv6 recv      77629    sent    77645   drop    0       forwarded       0
IPv6 ND recv   983      sent    995     drop    0
IPv6 MLD recv  0        sent    3       drop    0
IP vhlerr      0        hblener 0       lblener 0
IP fragerr     0        chkerr  0       protoer 0
UDP recv       76646    sent    76646   drop    0
UDP chkerr     0
Bytes received 10854602
Bytes sent     10824109
Processing err 11

fgrandel commented:
@cfriedt I think this one is ready for re-reviewing now.

kartben commented:
I'm obviously late here, but shouldn't these changes be added to the release notes? Renaming the compatible and Kconfig options may break things for people.
Also, it would probably make sense to rename `ti,cc13xx-cc26xx-rtc.yaml` to `ti,cc13xx-cc26xx-rtc-timer.yaml`?

fgrandel commented:
> I'm obviously late here,

@kartben It is never too late. :-) See #60177 

Thanks for catching these obvious issues!
